testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java

```diff
@@ -2473,6 +2474,44 @@ private static void checkConcatWithSeparatorInMSSQL(SqlOperatorFixture f) {
     f.checkString("concat_ws('', '', '', '')", "", "VARCHAR(0) NOT NULL");
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-6446">[CALCITE-6446]
+   * Add CONCAT_WS function (enabled in Spark library)</a>. */
+  @Test void testConcatWSFuncNullable() {
+    final SqlOperatorFixture f = fixture()
+        .setFor(SqlLibraryOperators.CONCAT_WS_SPARK)
+        .withLibrary(SqlLibrary.SPARK);
+    f.checkString("concat_ws(',', 'a')", "a", "VARCHAR(1) NOT NULL");
+    f.checkString("concat_ws(',', 'a', 'b', null, 'c')", "a,b,c",
+        "VARCHAR NOT NULL");
+    f.checkString("concat_ws(',', cast('a' as varchar), cast('b' as varchar))",
+        "a,b", "VARCHAR NOT NULL");
+    f.checkString("concat_ws(',', cast('a' as varchar(2)), cast('b' as varchar(1)))",
+        "a,b", "VARCHAR(4) NOT NULL");
+    f.checkString("concat_ws(',', '', '', '')", ",,", "VARCHAR(2) NOT NULL");
+    f.checkString("concat_ws(',', null, null, null)", "", "VARCHAR NOT NULL");
+    // returns null if the separator is null
+    f.checkNull("concat_ws(null, 'a', 'b')");
+    f.checkNull("concat_ws(null, null, null)");
+    f.checkString("concat_ws(',')", "", "VARCHAR(0) NOT NULL");
+    // if the separator is empty string, it's equivalent to CONCAT
+    f.checkString("concat_ws('', cast('a' as varchar(2)), cast('b' as varchar(1)))",
+        "ab", "VARCHAR(3) NOT NULL");
+    f.checkString("concat_ws('', '', '', '')", "", "VARCHAR(0) NOT NULL");
+    f.checkString("concat_ws(',', array['a', 'b'])", "a,b", "VARCHAR NOT NULL");
+    f.checkString("concat_ws(',', null)", "", "VARCHAR NOT NULL");
+    f.checkString("concat_ws(',', array['a', null, 'b'])", "a,b", "VARCHAR NOT NULL");
+    f.checkString("concat_ws(',', 1, 'b')", "1,b", "VARCHAR NOT NULL");
+    f.checkString("concat_ws(',', 100.0, 'b')", "100.0,b", "VARCHAR NOT NULL");
+    f.checkString("concat_ws(',', 'a', 'b', array['c'])", "a,b,c", "VARCHAR NOT NULL");
+    f.checkString("concat_ws(',', 'a', 'b', array['c'], array['d'])", "a,b,c,d",
+        "VARCHAR NOT NULL");
+    f.checkFails("^concat_ws(',', 100, 'b', array['c'])^",
```

Comment:
OperandTypeChecker has been updated.
Based on type coercion, `concat_ws(',', 100, 'b', array['c'])` can be supported, `100` is a basic type of `Integer` and can be cast to `String` type.   `concat_ws(',', 'a', 'b', array[100])`  is not supported yet because Calcite currently does not support type coercion within `array`.

Additionally, converting non-string arguments to strings is trivial in Calcite UDF.  Secondly, Spark has richer implicit conversions, and the above two cases can be run in Spark.